### DISCUSSION TITLE
WildFly OIDC adapter deprecation

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#9815] [Adapter deprecation]
Keycloak is deprecating most adapters in the future, for more information see https://www.keycloak.org/2023/03/adapter-deprecation-update.html.

WildFly 25 introduced native support for OpenID Connect with all the features from the Keycloak adapter and more. Migration to the WildFly native OpenID Connect is very easy as the WildFly team has taken great care to make this as simple as a move as possible.

Check out this [blog post](https://wildfly-security.github.io/wildfly-elytron/blog/securing-wildfly-apps-openid-connect/) from Farah Juma for more details.

We are not planning to support WildFly 25 and beyond with the Keycloak WildFly adapter.

**Comment by yeroc** [comment]
I read through the information at both links but didn't find anything related to how multi-tenancy is handled.  Is that something that is still supported or will we need to deploy separate application instances for each customer?

**Reply by stianst** [reply]
I believe WildFly team is planning to add support for multi-tenancy, or has already added. I'd suggest asking on WildFly forums though as that's the best place to request this being added.

**Reply by stianst** [reply]
WildFly team has an example for multi-tenancy support, see:
https://github.com/wildfly-security-incubator/elytron-examples/tree/master/multi-tenancy-oidc

**Reply by yeroc** [reply]
Thanks.  I couldn't find anything in the documentation.

**Comment by marcelnem** [comment]
@stianst  

For me, it is a bit confusing to understand what exactly is getting deprecated.  
Currently, there are many good examples of how to achieve fine-grained authorization (resource-based, scope-based, or completely dynamic and custom programmed) for Wildfly/Spring :
![image](https://user-images.githubusercontent.com/6547340/160350712-db226c6e-bdd4-45c0-8132-252ab9687269.png)

Current java / node js clients have functions such as "enforce()". It is possible to set scopes and claims in the access tokens which keycloak can then evaluate and grant/deny. 

To keep the authorization logic in Keycloak  (resources, scopes, policies, permissions), we need to map the endpoints of the (java/node/spring) REST API to the scope, claims, and resources in keycloak access tokens, which then keycloak grants or denies. 

The existing Wildfly client can also directly ask Keycloak for an augmented access token, if a scope or resource is missing in the access token. So the client (webapp) does not need to fetch new access tokens from keycloak each time before accessing a different REST API response. I wonder if this is possible with other open-source OICD clients. As this makes debugging and testing webapps much easier (there are many less 403 in webapp logs due to lacking permissions, because wildfly keycloak adapter, gets the new permissions automatically). 

I found the mentioned keywords in quarkus.keycloak : 
![image](https://user-images.githubusercontent.com/6547340/160353113-6e72ba46-ee80-4be6-ab6b-6a4e0ae80486.png)

But I am not sure how to achieve such fine-grained authorization with other OICD clients. 

In general, the move feels like starting all over again. All the community knowledge about integrations will be not applicable anymore and we will have to rewrite our webapp/db integrations for different technologies to get the same functionality we had already. This means costs without many benefits which usually management is not happy about. I can see that this will also be a moment to reevaluate the technology stack for companies. "If we have to rewrite everything anyway, let's reevaluate if we can do it easier with other tools". This can be partially mitigated by migration guides and quickstart-examples for the new technologies.

**Reply by stianst** [reply]
Around fine-grained authorization we will most likely provide a generic Java client that can be leveraged with different containers, and relatively easily be plugged in with a servlet-filter or something like that.

I appreciate your comment that this involves work on your end. However, putting it simply we do not have the capacity to both further develop the Keycloak server as well as adapters, and specifically for the WildFly adapter now that WildFly has native OIDC support it makes no sense for us to continue maintaining ours.

**Reply by stianst** [reply]
For authorization services specifically let's move the conversations around that to https://github.com/keycloak/keycloak/discussions/11681

**Reply by yeroc** [reply]
In theory a JASPIC module would allow for a nice level of integration with different containers.  I know older containers had significant bugs and inconsistencies but that seems to have gotten better [over the years](https://arjan-tijms.omnifaces.org/2016/12/the-state-of-portable-authentication-in.html) thanks to Arjan Tijms' work to bring more visibility to the issue.

**Comment by maimgrn** [comment]
Hello folks,

what about the "keycloak-servlet-filter-adapter" , 
will this also no longer supported in the near future?

This one:
<groupId>org.keycloak</groupId>
<artifactId>keycloak-servlet-filter-adapter</artifactId>

Thanks for the information

**Reply by stianst** [reply]
Yes, the OIDC servlet filter is also deprecated

**Reply by kkcmadhu** [reply]
> what about the "keycloak-servlet-filter-adapter" ,

I have started a seperate discussion on servlet filter adapter here. https://github.com/keycloak/keycloak/discussions/11790

**Comment by stianst** [comment]
Please take a look at https://www.keycloak.org/2023/03/adapter-deprecation-update.html for the updates to our plans our adapters. In summary for WildFly we are planning on delivering an authorization client library that can be used in combination with Elytron OIDC or Jakarta Security 3.0.
